### ISSUE TITLE
Calling unreference on one object appears to temporarily cause different object to be <Object#null>

### ISSUE THREAD

**Issue by bitwes** [godotengine/godot#69411]
### Godot version

4.0.beta5.official.89a33d28f

### System information

OSX 10.15.7

### Issue description

**TL;DR**:  Calling `unreference` on a function's local variable appears to cause a value in a dictionary, in a completely different part of the project, to be `<Object#null>`.  This happens less often if I `print` more.

I believe I have found an obscure race condition when calling `unreference`?  I have tried to simplify the steps to reproduce with no luck.   This occurs in [GUT](https://github.com/bitwes/Gut). 

When running the tests for GUT through GUT (using the stones to destroy the stones) I occasionally get a null object in a dictionary after calling `unreference` on a completely unrelated object.  This happens less frequently when there are more calls to `print` between the call to `unreference` and accessing the dictionary.

In the code below I have added debug `print` calls.  When these calls are present the error is less frequent.  Without the `print` calls and the logging for GUT set to 0, the error happens regularly.

## The call to `unreference`
This test function is making a local variable and then freeing it (it tests the implementation of `func unreference()` in `script_parser.gd`).  The scope of the unreferenced variable should be this function only.
``` gdscript
# res://test/unit/test_script_parser.gd
	func test_parsing_native_does_not_generate_orphans():
		var parsed = ParsedScript.new(Node2D)
		print('parsed = ', parsed)
		parsed.unreference()
		parsed = null
		assert_no_new_orphans()
```
This function is run dynamically through `gut.gd`.  `gut.gd` gets the names of methods and then does a ton of stuff to eventually call this method.

## The dictionary that gets a null object sometimes
`gut.gd` has an instance of `summary.gd`.  `summary.gd` has a dictionary that contains information about each test function that has been executed.  The dictionary is keyed with the name of the test function.  The value is an instance of an inner-class named `Test`.  `Test` class is defined in `summary.gd`.  I could explain more, but the call stacks in the output do a better job.

The below function handles adding and retrieval of values in the dictionary.  Occasionally, the value in the dictionary in `summary.gd` will be null after the call to `parsed.unreference` is made.

``` gdscript

# res://addons/gut/summary.gd.TestScript
class TestScript:
	var _tests = {}
	var _test_order = []

	...

	func get_test_obj(obj_name):
		print('--- get_test_obj stack ---')
		print_stack()
		print('---')
		if(!_tests.has(obj_name)):
			print('_tests does not have "', obj_name, '".')
			var to_add = Test.new()
			_tests[obj_name] = to_add
			_test_order.append(obj_name)
			print('    added "', obj_name, '" to _tests as ', to_add, '.  _tests[', obj_name, '] = ', _tests[obj_name])
		else:
			pass
			print('_tests has "', obj_name, '" key.  _tests[', obj_name, '] = ', _tests[obj_name])

		var to_return = _tests[obj_name]
		if(to_return == null):
			print('Return value is null.  _tests[', obj_name, '] = ', _tests[obj_name])
```

Commented and simplified output for when error occurs.  Here you can see
* the instance of `ParsedScript` is not the same instance added to the dictionary.
* `get_test_obj` adds a new instance of `Test` to the `_tests` dictionary for `"test_parsing_native_does_not_generate_orphans"` 
* Later `get_test_obj` sees that the key exists but the value is `<Object#null>`.
```
# the value of parsed in the test test_parsing_native_does_not_generate_orphans
parsed = <RefCounted#-9223371999273810438>

# The first call to get_test_obj 
# This call is triggered by assert_no_new_orphans
_tests does not have "test_parsing_native_does_not_generate_orphans".  
    added "test_parsing_native_does_not_generate_orphans" to _tests as <RefCounted#-9223371994525858310>.  _tests[test_parsing_native_does_not_generate_orphans] = <RefCounted#-9223371994525858310>

# The second call to get_test_obj.
# This call triggered when test_parsing_native_does_not_generate_orphans
# finishes.
_tests has "test_parsing_native_does_not_generate_orphans" key.  _tests[test_parsing_native_does_not_generate_orphans] = <Object#null>
Return value is null.  _tests[test_parsing_native_does_not_generate_orphans] = <Object#null>
```

Here is the output with the output of the `print_stack` call inside `get_test_obj`:
```
res://test/unit/test_script_parser.gd.TestParsedScript
parsed = <RefCounted#-9223371999273810438>
--- get_test_obj stack ---
Frame 0 - :113 in function 'get_test_obj'
Frame 1 - :132 in function 'add_pass'
Frame 2 - res://addons/gut/summary.gd:170 in function 'add_pass'
Frame 3 - res://addons/gut/gut.gd:1060 in function '_pass'
Frame 4 - res://addons/gut/test.gd:113 in function '_pass'
Frame 5 - res://addons/gut/test.gd:970 in function 'assert_no_new_orphans'
Frame 6 - :206 in function 'test_parsing_native_does_not_generate_orphans'
Frame 7 - res://addons/gut/gut.gd:794 in function '_run_test'
Frame 8 - res://addons/gut/gut.gd:999 in function '_test_the_scripts'
Frame 9 - res://addons/gut/gut.gd:1200 in function 'test_scripts'
Frame 10 - res://addons/gut/gui/GutRunner.gd:55 in function 'run_tests'
Frame 11 - res://addons/gut/gut_cmdln.gd:280 in function '_run_gut'
Frame 12 - res://addons/gut/gut_cmdln.gd:319 in function '_init'
---
_tests does not have "test_parsing_native_does_not_generate_orphans".
    added "test_parsing_native_does_not_generate_orphans" to _tests as <RefCounted#-9223371994525858310>.  _tests[test_parsing_native_does_not_generate_orphans] = <RefCounted#-9223371994525858310>
--- get_test_obj stack ---
Frame 0 - :113 in function 'get_test_obj'
Frame 1 - res://addons/gut/summary.gd:167 in function 'add_test'
Frame 2 - res://addons/gut/gut.gd:800 in function '_run_test'
Frame 3 - res://addons/gut/gut.gd:999 in function '_test_the_scripts'
Frame 4 - res://addons/gut/gut.gd:1200 in function 'test_scripts'
Frame 5 - res://addons/gut/gui/GutRunner.gd:55 in function 'run_tests'
Frame 6 - res://addons/gut/gut_cmdln.gd:280 in function '_run_gut'
Frame 7 - res://addons/gut/gut_cmdln.gd:319 in function '_init'
---
_tests has "test_parsing_native_does_not_generate_orphans" key.  _tests[test_parsing_native_does_not_generate_orphans] = <Object#null>
Return value is null.  _tests[test_parsing_native_does_not_generate_orphans] = <Object#null>
```

Let me know what other information you need.  This felt like a lot of information and also not enough.

## Get GUT
To get GUT and all of its tests you have to clone the Godot 4 branch (the zip only contains addons)
```
git clone -b godot_4 https://github.com/bitwes/Gut
```

## Run test script to get error
Running GUT with the following command should cause the error below to happen after a couple runs (occured 1st and 3rd time I ran with these steps).

```
<path to godot> -s res://addons/gut/gut_cmdln.gd  -d  --headless  -gselect=test_script_parser.gd -glog=0
```
The error:
```
Debugger Break, Reason: 'Invalid set index 'orphans' (on base: 'null instance') with value of type 'float'.'
*Frame 0 - res://addons/gut/gut.gd:827 in function '_run_test'
Enter "help" for assistance.
debug> bt
*Frame 0 - res://addons/gut/gut.gd:827 in function '_run_test'
 Frame 1 - res://addons/gut/gut.gd:991 in function '_test_the_scripts'
debug>
```

## More output = less often
Running GUT with a higher log level makes the error occur less frequently.
```
<path to godot> -s res://addons/gut/gut_cmdln.gd  -d  --headless  -gselect=test_script_parser.gd -glog=3
```

### Minimal reproduction project

As I said earlier, I was unable to simplify the steps to reproduce.  To get GUT and all of its tests you have to clone the Godot 4 branch (the zip only contains addons)
```
git clone -b godot_4 https://github.com/bitwes/Gut
```

**Comment by bitwes** [comment]
Occurs in Beta 6 `4.0.beta6.official.7f8ecffa5` as well.

**Comment by bitwes** [comment]
In Beta 6 (did not test in Beta 5), adding `await get_tree().process_frame` at different places in `test_parsing_native_does_not_generate_orphans` had different results.

## After instantiation seemed to fix it (workaround)
Adding `await get_tree().process_frame` after `var parsed = ParsedScript.new(Node2D)` appears to fix things.  I ran it about 20 times and no errors occurred.

This is a valid workaround.  There can't be many valid use cases for calling `unreference` on an object immediately after instantiating it.
```gdscript
func test_parsing_native_does_not_generate_orphans():
	var parsed = ParsedScript.new(Node2D)
	await get_tree().process_frame
	print('parsed = ', parsed)
	parsed.unreference()
	parsed = null
	assert_no_new_orphans()
```

## At start of method caused ocassional engine errors
If I add `await get_tree().process_frame` as the first line then occasionally I get a `malloc` error, and I got a program crash once.  Out of about 20 runs I got an error about 5 times and one crash.  These errors occurred **at the end of the run**, and the initial issue reported (dictionary value being null) never occurred.

```gdscript
func test_parsing_native_does_not_generate_orphans():
	await get_tree().process_frame
	var parsed = ParsedScript.new(Node2D)
	print('parsed = ', parsed)
	parsed.unreference()
	parsed = null
	assert_no_new_orphans()
```
There were two distinct errors I noticed and one crash.  `Ran Scripts matching "test_script_parser.gd"` is the last thing that GUT prints before it quits.

Here are two examples of "modified after being freed" error.  This error happened the most often.
```
Ran Scripts matching "test_script_parser.gd"
Godot(39172,0x10aac1dc0) malloc: Incorrect checksum for freed object 0x7fc8a08fbe38: probably modified after being freed.
Corrupt value: 0x0
Godot(39172,0x10aac1dc0) malloc: *** set a breakpoint in malloc_error_break to debug
[1]    39172 abort      "/Applications/Godot_4.0_b6.app/Contents/MacOS/Godot" -s  -d --headless 


Ran Scripts matching "test_script_parser.gd"
Godot(39476,0x111b28dc0) malloc: Incorrect checksum for freed object 0x7fe354afb160: probably modified after being freed.
Corrupt value: 0x600007fe354afcff
Godot(39476,0x111b28dc0) malloc: *** set a breakpoint in malloc_error_break to debug
[1]    39476 abort      "/Applications/Godot_4.0_b6.app/Contents/MacOS/Godot" -s  -d --headless 
```

Here is the other error, "pointer being freed was not allocated"
```
Ran Scripts matching "test_script_parser.gd"
Godot(39457,0x1183f4dc0) malloc: *** error for object 0x7fc2a85d3fa000: pointer being freed was not allocated
Godot(39457,0x1183f4dc0) malloc: *** set a breakpoint in malloc_error_break to debug
[1]    39457 abort      "/Applications/Godot_4.0_b6.app/Contents/MacOS/Godot" -s  -d --headless 
```

Here is the dump that occurred once.
```
================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.0.beta6.official (7f8ecffa56834dce3ccbd736738b613d51133dea)
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
[1] 1   libsystem_platform.dylib            0x00007fff6ddc85fd _sigtramp + 29
[2] 2   ???                                 0x0000000000000004 0x0 + 4
[3] JSON::get_data() const
[4] JSON::get_data() const
[5] JSON::get_data() const
[6] JSON::get_data() const
[7] Object* ClassDB::creator<ENetConnection>()
[8] Object::can_translate_messages() const
[9] RendererCompositorRD::_create_current()
[10] JSON::get_data() const
[11] JSON::get_data() const
[12] JSON::get_data() const
[13] JSON::get_data() const
[14] Object* ClassDB::creator<ENetConnection>()
[15] Object::can_translate_messages() const
[16] RendererCompositorRD::_create_current()
[17] JSON::get_data() const
[18] Object* ClassDB::creator<ENetConnection>()
[19] Object::can_translate_messages() const
[20] MultiplayerAPI::is_server()
[21] Object::can_translate_messages() const
[22] MultiplayerAPI::is_server()
[23] Object::can_translate_messages() const
[24] MultiplayerAPI::is_server()
[25] Object::can_translate_messages() const
[26] SceneTree::get_root() const
[27] 27  Godot                               0x00000001086b9fa3 Godot + 4640675
[28] RendererCompositorRD::_create_current()
[29] 29  libdyld.dylib                       0x00007fff6dbcfcc9 start + 1
-- END OF BACKTRACE --
================================================================
[1]    38376 abort      "/Applications/Godot_4.0_b6.app/Contents/MacOS/Godot" -s  -d --headless 
```
